pytorch_lightning/metrics/classification/auroc.py

```diff
@@ -0,0 +1,163 @@
+
+    Example (binary case):
+
+        >>> preds = torch.tensor([0.13, 0.26, 0.08, 0.19, 0.34])
+        >>> target = torch.tensor([0, 0, 1, 1, 1])
+        >>> auroc = AUROC(pos_label=1)
+        >>> auroc(preds, target)
+        tensor(0.5000)
+
+    Example (multiclass case):
+
+        >>> preds = torch.tensor([[0.90, 0.05, 0.05],
+        ...                       [0.05, 0.90, 0.05],
+        ...                       [0.05, 0.05, 0.90],
+        ...                       [0.85, 0.05, 0.10],
+        ...                       [0.10, 0.10, 0.80]])
+        >>> target = torch.tensor([0, 1, 1, 2, 2])
+        >>> auroc = AUROC(num_classes=3)
+        >>> auroc(preds, target)
+        tensor(0.7778)
+
+    """
+    def __init__(
+        self,
+        num_classes: Optional[int] = None,
+        pos_label: Optional[int] = None,
+        average: Optional[str] = 'macro',
+        max_fpr: Optional[float] = None,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Callable = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.num_classes = num_classes
+        self.pos_label = pos_label
+        self.average = average
+        self.max_fpr = max_fpr
+
+        allowed_average = (None, 'macro', 'weighted')
+        if self.average not in allowed_average:
+            raise ValueError('Argument `average` expected to be one of the following:'
+                             f' {allowed_average} but got {average}')
+
+        if self.max_fpr is not None:
+            if (not isinstance(max_fpr, float) and 0 < max_fpr <= 1):
+                raise ValueError(f"`max_fpr` should be a float in range (0, 1], got: {max_fpr}")
+
+            if LooseVersion(torch.__version__) < LooseVersion('1.6.0'):
+                raise RuntimeError(
+                    '`max_fpr` argument requires `torch.bucketize` which is not available below PyTorch version 1.6'
+                )
+
+        self.mode = None
+        self.add_state("preds", default=[], dist_reduce_fx=None)
+        self.add_state("target", default=[], dist_reduce_fx=None)
+
+        rank_zero_warn(
+            'Metric `AUROC` will save all targets and predictions in buffer.'
+            ' For large datasets this may lead to large memory footprint.'
+        )
+
+    def update(self, preds: torch.Tensor, target: torch.Tensor):
+        """
+        Update state with predictions and targets.
+
+        Args:
+            preds: Predictions from model (probabilities, or labels)
+            target: Ground truth labels
+        """
+        preds, target, mode = _auroc_update(preds, target)
+
+        self.preds.append(preds)
+        self.target.append(target)
+        self.mode = mode
```

Comment:
Should the mode be saved in a list too ?

Reply:
Thanks for the suggestion. So instead of saving it to a list, I instead included a check that `self.mode==mode`, which makes sure that the mode of the data (binary, multi-class, multi-label) did not change between batches (which it really should not).